### PR TITLE
install.sh patches gfortran 8.1.0 before building

### PR DESCRIPTION
**Pull request by rouson**
[links]:#
[contributing guidelines]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[issue]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[PR response img]: https://img.shields.io/issuestats/p/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square
[coverage]: https://img.shields.io/codecov/c/github/sourceryinstitute/OpenCoarrays/master.svg?style=flat-square

|  Avg response time                |  coverage on master         |
|:---------------------------------:|:---------------------------:|
| ![Issue Stats][PR response img]   | ![Codecov branch][coverage] |

## Summary of changes ##

Edited installer to detect when it is building GCC 8.1.0 and, if so, to patch the gfortran source.

## Rationale for changes ##

A GCC commit introduced just before the release of 8.1.0 caused regressions in OpenCoarrays.  The patch posted [here](https://gcc.gnu.org/ml/fortran/2018-04/msg00120.html) eliminates the regression by patching GCC 8.1.0 when installed by `install.sh`.

## Additional info and certifications ##

This pull request (PR) is a:

- [X] Bug fix
- [ ] Feature addition
- [ ] Other, Please describe:

### I certify that ###

- [X] I reviewed and followed the [contributing guidelines], including
      - Increasing test coverage for all feature-addition PRs
      - Increasing test coverage for all bug-fix PRs for which there
        does not already exist a related test that failed before the PR
      - At least maintaining test coverage for all other PRs
      - Ensuring that all tests pass when run locally 
      - Naming PR  to indicate work in progress (WIP) and to attach the PR
        to the appropriate bug report or feature request [issue]
      - White space (no trailing white space or white space errors may
        be introduced)
      - Commenting code where it is non-obvious and non-trivial
      - Logically atomic, self consistent and coherent commits
      - Commit message content
      - Waiting 24 hours before self-approving the PR to give another
        OpenCoarrays developer a chance to review my proposed code

### CODE REVIEW

**Comment by rouson** [comment]
@zbeekman It appears that the failing tests are only on macOS.  Is this related to issue #527?  Because this patch only addresses building GCC 8.1.0, I'd like to merge this and produce a 2.1.0 release ASAP for those interested in using OpenCoarrays with GCC 8.1.0.  Based on past experience, it's likely to be multiple months before 8.2.0 releases so patching 8.1.0 seems to be the best option for now.

**Comment by zbeekman** [comment]
@rouson It is related to #527 in the sense that GCC 8 breaks simple coarray programs...

But this is really thanks to the fact that custom MPICH bottles (and any MPICH or OpenMPI we get directly from Homebrew) specifies "gcc" as a dependency. I need to update and upload MPICH bottles that specify gcc@7 and gcc@6 as the dependency so that they won't link against (and install) gcc 8. If the .so version of libgfortran (and others?) hadn't been bumped by GCC 8, this *might* not have been an issue, but at any rate I should update the bottles.

TL;DR;
-------

Yes, this is purely a CI issue on mac. I started to workup a fix but I'm not all the way there yet. For the time being, this build failure is completely independent of the actual changes so

LGTM

File: prerequisites/build-functions/build_and_install.sh [hunk 1]
```diff
@@ -45,13 +44,23 @@ build_and_install()
...
-
     info "pushd ${download_path}/${package_source_directory} "
     pushd "${download_path}/${package_source_directory}"
 
-    # Switch download mechanism, if wget is not available
+    # Patch gfortran if necessary
+    export patches_dir="${OPENCOARRAYS_SRC_DIR}/prerequisites/build-functions/patches/${package_to_build}/${version_to_build}"
+    if [[ -d "${patches_dir:-}" ]]; then
+      #TODO: check whether multiple patch files exist in $patches_dir and apply each one
+      export patch_file="$(ls "$patches_dir")"
```
Comment (zbeekman) [inline prerequisites/build-functions/build_and_install.sh]: `ls` is not designed for the output to be consumed programatically. Applying all patches using shell globbing is probably more robust and solves the `#TODO` commend. I'll push a commit to do this. This will be good to use to apply Andre's patch to MPICH too.

**Comment by zbeekman** [comment]
@rouson Let's make sure the InstallScript tests pass on Linux and mac, and then merge this once we have confirmed that they do. (I pushed a commit to apply all the patches in the patch directory.)

File: prerequisites/build-functions/patches/gcc/8.1.0/partially-revert-rev259385.diff [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+@@ -1860,7 +1860,7 @@ conv_caf_send (gfc_code *code) {
+ 
+   lhs_expr = code->ext.actual->expr;
+   rhs_expr = code->ext.actual->next->expr;
+-  may_require_tmp = gfc_check_dependency (lhs_expr, rhs_expr, false) == 0
++  may_require_tmp = gfc_check_dependency (lhs_expr, rhs_expr, true) == 0
+ 		    ? boolean_false_node : boolean_true_node;
+   gfc_init_block (&block);
+ 
+diff --git a/gcc/testsuite/gfortran.dg/coarray_dependency_1.f90 b/gcc/testsuite/gfortran.dg/coarray_dependency_1.f90
```
Comment (zbeekman) [inline prerequisites/build-functions/patches/gcc/8.1.0/partially-revert-rev259385.diff]: Unless we anticipate running the GCC test-suite, we may want to remove this line (L30) to the end off the patch. The less hunks we are patching the less likely the application will fail. Just an idea, take it or leave it.

Comment (rouson) [reply prerequisites/build-functions/patches/gcc/8.1.0/partially-revert-rev259385.diff]: @zbeekman Great idea. Feel free to push a commit that removes that part of the patch or let me know if I should do so.

**Comment by rouson** [comment]
It appears that the Linux builds passed except for the allowed failures.
LGTM
